c++/src/kj/io.c++

```diff
@@ -401,11 +417,47 @@ void FdOutputStream::write(ArrayPtr<const ArrayPtr<const byte>> pieces) {
   }
 
   while (current < iov.end()) {
+    size_t iovCount = iov.end() - current;
+
+#if __APPLE__
+    // MacOS will fail if you give it more than INT_MAX bytes to write at once. We can solve this
+    // by carefully truncating the list to a lesser number. Why the OS doesn't just return a short
+    // write itself, I don't know.
+    size_t totalSize = 0;
+    struct iovec* editedPiece = nullptr;
+    size_t editedPieceOriginalLen = 0;
+    for (auto i: kj::zeroTo(iovCount)) {
+      auto& piece = *(current + i);
+      totalSize += piece.iov_len;
+      if (totalSize >= WRITE_CLAMP_SIZE) {
+        // Truncate the list after this piece.
+        iovCount = i + 1;
+
+        if (totalSize > WRITE_CLAMP_SIZE) {
+          // We also have to truncate this piece. Patch it in-place and plan to fix it later.
+          editedPiece = &piece;
+          editedPieceOriginalLen = piece.iov_len;
+          size_t overage = totalSize - WRITE_CLAMP_SIZE;
+          piece.iov_len -= overage;
+        }
+
+        break;
+      }
+    }
+#endif
+
     // Issue the write.
     ssize_t n = 0;
-    KJ_SYSCALL(n = ::writev(fd, current, iov.end() - current), fd);
+    KJ_SYSCALL(n = ::writev(fd, current, iovCount), fd);
     KJ_ASSERT(n > 0, "writev() returned zero.");
 
+#if __APPLE__
+    // If we patched the list above, unpatch now.
+    if (editedPiece != nullptr) {
```

Comment:
What is the exception flow? Should we worry about it?

Reply:
The iov array was allocated on the stack earlier in this function, line 403. So on exception it harmlessly goes away.